### PR TITLE
Allow excluding folder names when scanning for dangling indices

### PR DESCRIPTION
**Pull request by ywelsch**
ES is scanning for dangling indices on every cluster state update. For this, it lists the subfolders of the indices directory to determine which extra index directories exist on the node where there's no corresponding index in the cluster state. These are potential targets for dangling index import. On certain machine types, and with large number of indices, this subfolder listing can be horribly slow. This means that every cluster state update will be slowed down by potentially hundreds of milliseconds. One of the reasons for this poor performance is that `Files.isDirectory()` is a relatively expensive call on some OS and JDK versions. There is no need though to do all these `isDirectory` calls for folders which we know we are going to discard anyhow in the next step of the dangling indices logic. This PR allows adding an exclusion predicate to the `availableIndexFolders` methods which can dramatically speed up this method when scanning for dangling indices. Benchmarks are below:

```
Linux:

Benchmark                                                     Mode  Cnt        Score       Error  Units
AvailableIndexFoldersBenchmark.availableIndexFolderNaive      avgt   15  6618587.094 ± 65112.212  ns/op
AvailableIndexFoldersBenchmark.availableIndexFolderOptimized  avgt   15  2654325.642 ± 11821.622  ns/op
```

```
Windows (virtualized):

Benchmark                                                     Mode  Cnt          Score         Error  Units
AvailableIndexFoldersBenchmark.availableIndexFolderNaive      avgt   15  112902689.566 ± 6225614.594  ns/op
AvailableIndexFoldersBenchmark.availableIndexFolderOptimized  avgt   15    2492549.300 ±  103180.702  ns/op
```

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/gateway/MetaStateService.java [hunk 1]
```diff
@@ -85,7 +85,7 @@ public IndexMetaData loadIndexState(Index index) throws IOException {
      */
     List<IndexMetaData> loadIndicesStates(Predicate<String> excludeIndexPathIdsPredicate) throws IOException {
         List<IndexMetaData> indexMetaDataList = new ArrayList<>();
-        for (String indexFolderName : nodeEnv.availableIndexFolders()) {
+        for (String indexFolderName : nodeEnv.availableIndexFolders(excludeIndexPathIdsPredicate)) {
             if (excludeIndexPathIdsPredicate.test(indexFolderName)) {
                 continue;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/MetaStateService.java]: I think something has gone badly wrong if we end up here. Can we at least `assert false`, and maybe throw something in prod code too?

**Review by DaveCTurner — APPROVED** [review]
45x speedup on Windows :yum: LGTM, I left one nit.

File: benchmarks/src/main/java/org/elasticsearch/benchmark/fs/AvailableIndexFoldersBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package org.elasticsearch.benchmark.fs;
```
Comment (DaveCTurner) [inline benchmarks/src/main/java/org/elasticsearch/benchmark/fs/AvailableIndexFoldersBenchmark.java]: Needs a licence, apparently.

**Comment by jasontedor** [comment]
I do not think the Windows benchmarks should be taken as gospel for the performance gains on Windows. I had done some benchmarking in the past on Windows in a virtualized environment, and later observed that the virtualized results were quite far from the results on a bare-metal Windows machine.

This is not an argument against this change, it is a good change, and the Linux benchmarks clearly demonstrate that. Only that we should be cautious about thinking how big the gains actually are on Windows.

File: benchmarks/build.gradle [hunk 3]
```diff
@@ -44,6 +44,8 @@ compileJava.options.compilerArgs << "-Xlint:-cast,-deprecation,-rawtypes,-try,-u
 // needs to be added separately otherwise Gradle will quote it and javac will fail
 compileJava.options.compilerArgs.addAll(["-processor", "org.openjdk.jmh.generators.BenchmarkProcessor"])
 
+run.executable = new File(project.runtimeJavaHome, 'bin/java')
```
Comment (jasontedor) [inline benchmarks/build.gradle]: Would you extract this to a separate change, and explaining why it's needed?

**Review by jasontedor — APPROVED** [review]
Good change. I only left a request for a unit test for the new functionality on `NodeEnvironment`.

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 4]
```diff
@@ -818,13 +819,21 @@ public int getNodeLockId() {
...
      */
     public Set<String> availableIndexFolders() throws IOException {
+        return availableIndexFolders(p -> false);
+    }
+
+    /**
+     * Returns folder names in ${data.paths}/nodes/{node.id}/indices folder that don't match the given predicate.
+     * @param excludeIndexPathIdsPredicate folder names to exclude
+     */
+    public Set<String> availableIndexFolders(Predicate<String> excludeIndexPathIdsPredicate) throws IOException {
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: added in 9663898

**Pull request merged**